File: crawler/match_crawler.py
"""
Match Crawler for KeucolaTV
Crawl live football match links using Selenium
"""

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
import json
import time
from datetime import datetime

class MatchCrawler:
    def __init__(self, headless=True):
        """Initialize crawler with Chrome driver"""
        self.base_url = "https://st.90phut27.com"
        self.matches_url = f"{self.base_url}/home"
        
        # Setup Chrome options
        chrome_options = Options()
        if headless:
            chrome_options.add_argument("--headless")
        chrome_options.add_argument("--no-sandbox")
        chrome_options.add_argument("--disable-dev-shm-usage")
        chrome_options.add_argument("--disable-gpu")
        chrome_options.add_argument("--window-size=1920,1080")
        chrome_options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
        
        # Initialize driver
        self.driver = webdriver.Chrome(options=chrome_options)
        self.wait = WebDriverWait(self.driver, 10)
    
    def crawl_matches(self):
        """Crawl all matches from the page"""
        try:
            print(f"🔍 Đang crawl matches từ {self.matches_url}...")
            
            # Navigate to page
            self.driver.get(self.matches_url)
            
            # Wait for page to load
            time.sleep(3)
            
            # Wait for match links to appear
            try:
                self.wait.until(
                    EC.presence_of_element_located((By.CSS_SELECTOR, 'a.match-link, a[href*="/truc-tiep/"]'))
                )
            except:
                print("⚠️ Không tìm thấy match links, có thể trang đang load...")
            
            # Find all match links (90phut27 uses /truc-tiep pattern)
            match_links = self.driver.find_elements(By.CSS_SELECTOR, 'a.match-link, a[href*="/truc-tiep/"]')
            
            matches = []
            seen_urls = set()
            
            for link in match_links:
                try:
                    href = link.get_attribute('href')
                    
                    # Skip if already processed or is main page
                    if not href or href in seen_urls or '/truc-tiep/' not in href:
                        continue
                    
                    seen_urls.add(href)
                    
                    # Parse URL - 90phut27 format: /truc-tiep/match-slug-id
                    path_parts = href.replace(self.base_url, '').split('/')
                    path_parts = [p for p in path_parts if p]
                    
                    # Extract match ID from URL
                    if len(path_parts) >= 2 and 'truc-tiep' in path_parts:
                        match_id = path_parts[-1]  # Last part is the match slug-id
                        stream_id = match_id  # Use same as match_id for 90phut27
                        
                        # Find parent container (div.xitem2)
                        try:
                            parent = link.find_element(By.XPATH, './ancestor::div[contains(@class, "xitem2")]')
                        except:
                            # Fallback to any parent div
                            try:
                                parent = link.find_element(By.XPATH, './parent::div')
                            except:
                                parent = link
                        
                        # Extract team names from span.team-home .name and span.team-away .name
                        home_team = ''
                        away_team = ''
                        try:
                            home_team_elem = parent.find_element(By.CSS_SELECTOR, 'span.team-home .name')
                            home_team = home_team_elem.text.strip()
                        except:
                            pass
                        
                        try:
                            away_team_elem = parent.find_element(By.CSS_SELECTOR, 'span.team-away .name')
                            away_team = away_team_elem.text.strip()
                        except:
                            pass
                        
                        # Extract match time from span.xtime b
                        match_time = ''
                        try:
                            time_elem = parent.find_element(By.CSS_SELECTOR, 'span.xtime b')
                            match_time = time_elem.text.strip()
                        except:
                            pass
                        
                        # Extract status from span.xstatus or span.match-status
                        status = 'Sắp diễn ra'
                        try:
                            status_elem = parent.find_element(By.CSS_SELECTOR, 'span.xstatus')
                            status_text = status_elem.text.strip()
                            if 'Live' in status_text or 'live' in status_text:
                                status = 'Đang diễn ra'
                            elif status_text:
                                status = status_text
                        except:
                            try:
                                status_elem = parent.find_element(By.CSS_SELECTOR, 'span.match-status')
                                status_text = status_elem.text.strip()
                                if status_text:
                                    status = 'Đang diễn ra'  # If match-status exists, it's usually live
                            except:
                                pass
                        
                        match_data = {
                            'match_id': match_id,
                            'stream_id': stream_id,
                            'url': href,
                            'home_team': home_team,
                            'away_team': away_team,
                            'match_time': match_time,
                            'full_text': f"{home_team} vs {away_team}",
                            'status': status,
                            'source': '90phut27',
                            'crawled_at': datetime.now().isoformat()
                        }
                        
                        matches.append(match_data)
                        
                except Exception as e:
                    print(f"⚠️ Lỗi khi parse link: {e}")
                    continue
            
            print(f"✅ Crawled {len(matches)} matches")
            return matches
            
        except Exception as e:
            print(f"❌ Lỗi khi crawl: {e}")
            raise
    
    def crawl_match_detail(self, match_url):
        """Crawl detail of a specific match"""
        try:
            print(f"🔍 Crawling match detail: {match_url}")
            
            self.driver.get(match_url)
            time.sleep(3)
            
            # Find iframe sources (stream URLs)
            iframes = self.driver.find_elements(By.TAG_NAME, 'iframe')
            stream_urls = []
            
            for iframe in iframes:
                src = iframe.get_attribute('src')
                if src:
                    stream_urls.append(src)
            
            # Find video sources
            videos = self.driver.find_elements(By.TAG_NAME, 'video')
            for video in videos:
                sources = video.find_elements(By.TAG_NAME, 'source')
                for source in sources:
                    src = source.get_attribute('src')
                    if src:
                        stream_urls.append(src)
            
            return {
                'url': match_url,
                'stream_urls': stream_urls,
                'crawled_at': datetime.now().isoformat()
            }
            
        except Exception as e:
            print(f"❌ Lỗi khi crawl match detail: {e}")
            raise
    
    def save_to_json(self, matches, filename='matches.json'):
        """Save matches to JSON file"""
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(matches, f, ensure_ascii=False, indent=2)
        print(f"💾 Đã lưu {len(matches)} matches vào {filename}")
    
    def close(self):
        """Close browser"""
        if self.driver:
            self.driver.quit()
            print("🔒 Đã đóng browser")
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


def main():
    """Main function to test crawler"""
    print("🚀 Starting Match Crawler...\n")
    
    # Use context manager to auto-close browser
    with MatchCrawler(headless=False) as crawler:
        # Crawl matches
        matches = crawler.crawl_matches()
        
        # Print results
        print(f"\n📊 Kết quả crawl:")
        print(f"Tổng số trận: {len(matches)}\n")
        
        if matches:
            print("📋 Danh sách 5 trận đầu tiên:")
            for i, match in enumerate(matches[:5], 1):
                print(f"\n{i}. {match['home_team']} vs {match['away_team']}")
                print(f"   Thời gian: {match['match_time']}")
                print(f"   URL: {match['url']}")
                print(f"   Match ID: {match['match_id']}")
            
            # Save to JSON
            crawler.save_to_json(matches)
        else:
            print("⚠️ Không tìm thấy matches nào")
    
    print("\n✅ Crawler completed!")


if __name__ == "__main__":
    main()

File: backend/src/utils/matchCrawler.js
/**
 * Crawler để lấy link trực tiếp bóng đá từ KeucolaTV
 * File: backend/src/utils/matchCrawler.js
 */

import axios from "axios";
import * as cheerio from "cheerio";

const BASE_URL = "https://keucolatv.live";
const MATCHES_URL = `${BASE_URL}/truc-tiep-bong-da`;

/**
 * Crawl danh sách trận đấu từ KeucolaTV
 * @returns {Promise<Array>} Danh sách trận đấu
 */
export const crawlMatches = async () => {
  try {
    console.log("🔍 Đang crawl matches từ KeucolaTV...");

    // Fetch HTML
    const response = await axios.get(MATCHES_URL, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        Accept:
          "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "vi-VN,vi;q=0.9,en-US;q=0.8,en;q=0.7",
      },
      timeout: 10000,
    });

    const html = response.data;
    const $ = cheerio.load(html);
    const matches = [];

    // Tìm tất cả các match cards
    // Pattern: /truc-tiep-bong-da/{matchId}/{streamId}
    $('a[href^="/truc-tiep-bong-da/"]').each((index, element) => {
      const $el = $(element);
      const href = $el.attr("href");

      // Skip nếu là link chính (không có matchId)
      if (href === "/truc-tiep-bong-da" || href === "/truc-tiep-bong-da/") {
        return;
      }

      // Extract match info
      const fullUrl = `${BASE_URL}${href}`;
      const pathParts = href.split("/").filter((p) => p);

      if (pathParts.length >= 3) {
        const matchId = pathParts[1];
        const streamId = pathParts[2];

        // Lấy thông tin từ HTML
        const matchText = $el.text().trim();

        // Tìm parent container để lấy thêm info
        const $parent = $el.closest("div");
        const league = $parent
          .find('.league-name, [class*="league"]')
          .first()
          .text()
          .trim();
        const status = $parent
          .find('.status, [class*="status"]')
          .first()
          .text()
          .trim();

        // Lấy tên đội
        const teams = matchText.split("-").map((t) => t.trim());
        const homeTeam = teams[0] || "";
        const awayTeam = teams[1] || "";

        // Tìm thời gian
        const timeMatch = matchText.match(/(\d{1,2}:\d{2})/);
        const matchTime = timeMatch ? timeMatch[1] : "";

        matches.push({
          matchId,
          streamId,
          url: fullUrl,
          league: league || "Unknown League",
          homeTeam,
          awayTeam,
          matchTime,
          status: status || "Sắp diễn ra",
          source: "KeucolaTV",
          crawledAt: new Date(),
        });
      }
    });

    // Remove duplicates based on matchId
    const uniqueMatches = Array.from(
      new Map(matches.map((m) => [m.matchId, m])).values()
    );

    console.log(`✅ Crawled ${uniqueMatches.length} matches từ KeucolaTV`);
    return uniqueMatches;
  } catch (error) {
    console.error("❌ Lỗi khi crawl matches:", error.message);
    throw error;
  }
};

/**
 * Crawl chi tiết một trận đấu
 * @param {string} matchUrl - URL của trận đấu
 * @returns {Promise<Object>} Chi tiết trận đấu
 */
export const crawlMatchDetail = async (matchUrl) => {
  try {
    const response = await axios.get(matchUrl, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      },
      timeout: 10000,
    });

    const html = response.data;
    const $ = cheerio.load(html);

    // Extract stream URLs (iframe src, video src, etc.)
    const streamUrls = [];

    $("iframe").each((i, el) => {
      const src = $(el).attr("src");
      if (src) {
        streamUrls.push(src);
      }
    });

    $("video source").each((i, el) => {
      const src = $(el).attr("src");
      if (src) {
        streamUrls.push(src);
      }
    });

    return {
      url: matchUrl,
      streamUrls,
      crawledAt: new Date(),
    };
  } catch (error) {
    console.error("❌ Lỗi khi crawl match detail:", error.message);
    throw error;
  }
};

/**
 * Test crawler
 */
export const testCrawler = async () => {
  try {
    console.log("🧪 Testing crawler...\n");

    const matches = await crawlMatches();

    console.log("\n📊 Kết quả crawl:");
    console.log(`Tổng số trận: ${matches.length}\n`);

    if (matches.length > 0) {
      console.log("📋 Danh sách 5 trận đầu tiên:");
      matches.slice(0, 5).forEach((match, index) => {
        console.log(`\n${index + 1}. ${match.homeTeam} vs ${match.awayTeam}`);
        console.log(`   Giải đấu: ${match.league}`);
        console.log(`   Thời gian: ${match.matchTime}`);
        console.log(`   Trạng thái: ${match.status}`);
        console.log(`   URL: ${match.url}`);
        console.log(`   Match ID: ${match.matchId}`);
      });
    }

    return matches;
  } catch (error) {
    console.error("❌ Test failed:", error);
    throw error;
  }
};

export default {
  crawlMatches,
  crawlMatchDetail,
  testCrawler,
};
